### ISSUE TITLE
[Feature] Support var-type for defining types

### ISSUE THREAD

**Issue by manuth** [Microsoft/TypeScript#27261]
## Search Terms
"var type"
"dynamic type"

## Suggestion

I'm used to always define the types of all members and variables for a better readability which is why the reason why I always enable the tsling-rule which forces type-definitions.

But there are some situations where you can't define a type, for instance because its interface is not exported by a module.

For that reason I'd like to have an equivalent to `C#`s `var`-type added:

***C#:***
```cs
var foo = "bar";
foo = 1; // This causes an error because `foo` is a string.
```

***TypeScript:***
```ts
let foo: var = "bar";
foo = 1; // Jut like in `C#` this here should cause an error, too.
```

## Use Cases
So let's say I force every variable and member to have type-definitions and I have installed a module whose `.d.ts` file looks as followed:

```ts
export = CronParser
[...]
declare class CronExpression {
    [...]
}
declare namespace CronParser {
  /** Wrapper for CronExpression.parse method */
  function parseExpression(expression: string, options?: ParserOptions): CronExpression
  /** Parse crontab file */
  function parseFile(filePath: string, callback: (err: any, data: StringResult) => any): void
  /** Parse content string */
  function parseString(data: string): StringResult
}
```

If I try to use its features tslint will throw an error because the type of the variable is not defined:

```ts
import * as parser from "cron-parser";
[...]
let result = parser.parseExpression("0 0 1 * *"); // tslint: expected variable-declaration: 'x' to have a typedef
```

It would be pretty useful to add a type which tells typescript to automatically recognize its type (just like it normally would):

```ts
import * as parser from "cron-parser";
[...]
let result: var = parser.parseExpression("0 0 1 * *");
```

## Examples
```ts
import * as parser from "cron-parser";
[...]
let result: var = parser.parseExpression("0 0 1 * *");
```

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by dsherret** [comment]
@manuth c# doesn't have a "var type". It has a `var` keyword and the closest equivalent to that in TypeScript is the `let` keyword.

```ts
let foo = "bar";
foo = 1; // errors, just like in c#
```

Without the context of tslint, does it makes sense for the language to introduce something it already has?

**Comment by manuth** [comment]
Well currently I'm working around this error by temporarily disabling the rule which is pretty dirty though:
```ts
        // tslint:disable-next-line:typedef
```

The problem about disabling or relaxing the tslint-rule is that I'm not getting notified about variables and members where I have missed to define types anymore.

That's why I thought that adding such a "var"-type'd be the easiest way since it doesn't even have to do anything since TypeScript already does the same thing when no type is specified.

**Comment by andy-ms** [comment]
Duplicate of #26593

**Comment by RyanCavanaugh** [comment]
The real fix here is to not enable bad lint rules. Requiring a type annotation on every variable declaration is counterproductive at best.

**Comment by manuth** [comment]
Well actually I enabled it because I want to keep my code as consistent as possible which means that I either want to have definitions everywhere or definitions only at the places where it's required.

Sadly there is no option to *disallow* unnecessary type-definitions using tslint.

Do you think it'd be a good idea to open up an issue in the tslint repo about this?

**Comment by andy-ms** [comment]
Enabling `--noImplicitAny` will tell you if you forgot a type declaration where one is required.

**Comment by manuth** [comment]
That is true, yeah - I have this enabled, too.
What I'm planning to is to force to *not specify* the type if it can be resolved implicitly.
